src/webui/api/searchcontroller.cpp

```diff
@@ -0,0 +1,359 @@
+    else
+        setResult(getResults(searchResults, searchHandler->isActive(), size));
+}
+
+void SearchController::deleteAction()
+{
+    checkParams({"id"});
+
+    const QString sessionId = sessionManager()->session()->id();
+    const QString id = params()["id"].trimmed();
+
+    if (!m_searches.contains(sessionId))
+        throw APIError(APIErrorType::NotFound);
+
+    QHash<QString, SearchHandler *> &searches = m_searches[sessionId];
+
+    if (!searches.contains(id))
+        throw APIError(APIErrorType::NotFound);
+
+    SearchHandler *searchHandler = searches[id];
+    searchHandler->cancelSearch();
+    delete searches.take(id);
+}
+
+void SearchController::categoriesAction()
+{
+    QStringList categories;
+    const QString name = params()["pluginName"].trimmed();
+
+    categories << SearchPluginManager::categoryFullName("all");
+    for (const QString &category : SearchPluginManager::instance()->getPluginCategories(name))
+        categories << SearchPluginManager::categoryFullName(category);
+
+    setResult(categories.join(","));
+}
+
+void SearchController::pluginsAction()
+{
+    const QStringList allPlugins = SearchPluginManager::instance()->allPlugins();
+    setResult(getPluginsInfo(allPlugins));
+}
+
+void SearchController::installPluginAction()
+{
+    checkParams({"sources"});
+
+    const QStringList sources = params()["sources"].split('|');
+    for (const QString &source : sources)
+        SearchPluginManager::instance()->installPlugin(source);
+}
+
+void SearchController::uninstallPluginAction()
+{
+    checkParams({"names"});
+
+    const QStringList names = params()["names"].split('|');
+    for (const QString &name : names)
+        SearchPluginManager::instance()->uninstallPlugin(name.trimmed());
+}
+
+void SearchController::enablePluginAction()
+{
+    checkParams({"names", "enable"});
+
+    const QStringList names = params()["names"].split('|');
+    const bool enable = parseBool(params()["enable"].trimmed(), false);
+
+    for (const QString &name : names)
+        SearchPluginManager::instance()->enablePlugin(name.trimmed(), enable);
+}
+
+void SearchController::updatePluginsAction()
+{
+    SearchPluginManager *pluginManager = SearchPluginManager::instance();
+
+    connect(pluginManager, &SearchPluginManager::checkForUpdatesFinished, this, &SearchController::checkForUpdatesFinished);
+    connect(pluginManager, &SearchPluginManager::checkForUpdatesFailed, this, &SearchController::checkForUpdatesFailed);
+    pluginManager->checkForUpdates();
+}
+
+void SearchController::searchTimeout(const QString &sessionId, const QString &id)
+{
+    QHash<QString, SearchHandler *> &searches = m_searches[sessionId];
+
+    if (m_searches.contains(sessionId) && searches.contains(id))
+    {
+        Logger::instance()->addMessage(tr("Deleting search session %1 from web session %2 due to timeout").arg(id).arg(sessionId), Log::INFO);
```

Comment:
`.arg(id, sessionId)`  and please check other similar instances, combine multiple .arg() whenever possible.